### DISCUSSION TITLE
运行时配置中的动态路由代码中的patch是什么？

### DISCUSSION THREAD

**Discussion by qingfengmy** [umijs/umi#9456] [Q&A]
```
let extraRoutes;
 
export function patchClientRoutes({ routes }) {
  // 根据 extraRoutes 对 routes 做一些修改
  patch(routes, extraRoutes);
}
 
export function render(oldRender) {
  fetch('/api/routes')
    .then((res) => res.json())
    .then((res) => {
      extraRoutes = res.routes;
      oldRender();
    });
}
```
https://umijs.org/docs/api/runtime-config#patchclientroutes-routes-

patch(routes, extraRoutes);

里的patch是什么？

**Comment by fz6m** [comment]
根据你的需要在运行时修改路由，多见于动态路由的场景，先从服务端拉取配置，再根据配置里的权限来修改路由表。

**Reply by weilun0510** [reply]
> 根据你的需要在运行时修改路由，多见于动态路由的场景，先从服务端拉取配置，再根据配置里的权限来修改路由表。

他想问的可能是patch函数是从哪导入的。 
而我想问的是 `patchClientRoutes` 好像不起作用

```
// 动态更新路由
export function patchClientRoutes({ routes }: any) {
  console.log('routes: ', routes);
  routes.unshift({
    path: '/foo',
    element: <div>2</div>,
  });
}
```

没见到打印`routes`，请问怎么回事

**Reply by fz6m** [reply]
给一个最小复现看看。

**Reply by weilun0510** [reply]
> 给一个最小复现看看。

https://github.com/weilun0510/umi-dome-patchClientRoutes

我好像发现问题了，去掉 `export default {}` 后就能找到`/foo`，这是什么原因？我在`export default {}`里配置了一些信息

**Reply by fz6m** [reply]
现在的 `app.ts` 不支持 `default` 导出和 独立导出 混用，只能取其一。

可以等 #9708 解决该问题。

**Reply by weilun0510** [reply]
> 现在的 `app.ts` 不支持 `default` 导出和 独立导出 混用，只能取其一。
> 
> 可以等 #9708 解决该问题。

好的。但是独立导出`getInitialState`函数可以和`default`导出混用，有点奇怪

**Reply by weilun0510** [reply]
请问这个问题解决了吗？
我需要在appt.tsx中配置 request，`export default {
  request: requestConfig,
  layout: {
    // 自定义 403 页面
    noAccessible: <>无权限展示</>,
    pure: true, // 删除掉layout所有的自带界面
  },
};` 有需要使用patchClientRoutes动态更新路由。
